### PR TITLE
Discovery indexing: Ensuring discovery configuration is used during indexing

### PR DESCRIPTION
**Pull request by benbosman**
We noticed a few problems with the discovery indexing:
* If you create a configuration, link it to a collection, but don't add `<property name="id" value="somevalue"/>` where somevalue is unique per configuration, it will be ignored.
* It's not possible to create a discovery configuration applied to items without linking it to collections. This is however a useful feature to search with a custom discovery configuration and display facets specific to that configuration

This has been solved by:
* Making sure the list of configurations is a HashSet which avoids duplicate configurations (instead of a HashMap which only avoids duplicate ids and only supports one configuration without an id)
* Creating an `indexAlways` property which states the configuration should always be included when indexing items

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoveryConfiguration.java [hunk 1]
```diff
@@ -51,6 +51,15 @@
     private DiscoveryHitHighlightingConfiguration hitHighlightingConfiguration;
     private DiscoveryMoreLikeThisConfiguration moreLikeThisConfiguration;
     private boolean spellCheckEnabled;
+    private boolean indexAlways = false;
+
+    public boolean isIndexAlways() {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoveryConfiguration.java]: Please add Javadocs.  I'm not sure I understand this configuration and when something would be "true" vs "false".  Does "false" mean something is *not* indexed? or only indexed occasionally?  I think we should have more specific examples of usage in the Javadocs here -- I'm having difficulty understanding this property myself and why it's needed only on Entities configurations, and not on others (like `workflowConfiguration`, `workspaceConfiguration`, etc)

**Review by tdonohue — COMMENTED** [review]
@benbosman : I'm a little confused by the new `indexAlways` configuration here, as I don't quite understand its purpose. It might be that the name is confusing me, but more description/explanation would definitely help me out (see below)

File: dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoveryConfiguration.java (hunk 1 above)
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoveryConfiguration.java]: The Javadocs have been included.
The `workflowConfiguration` and `workspaceConfiguration` currently use custom code in the SearchUtils and SolrServiceImpl class to override the behavior. Indexing workspace and workflow items is still planned to be improved
This PR is focussed on entities configurations, and avoids creating custom code in SearchUtils and SolrServiceImpl with a hard-coded list of potential entity types

**Review by paulo-graca — APPROVED** [review]
This PR seems ok to me. I've also test it and didn't found any issue.

File: dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoveryConfigurationService.java [hunk 2]
```diff
@@ -75,6 +76,17 @@ public DiscoveryConfiguration getDiscoveryConfigurationByNameOrDso(final String
         }
     }
 
+    public List<DiscoveryConfiguration> getIndexAlwaysConfigurations() {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoveryConfigurationService.java]: Missing Javadocs.  This is a straightforward method, but we should get in the habit of adding Javadocs everywhere.

**Review by tdonohue — APPROVED** [review]
👍 This looks good overall.  However, I noted one public method here that is missing Javadocs.  I'd appreciate them being added before merging.  Beyond that, this looks fine to me.

File: dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoveryConfigurationService.java (hunk 2 above)
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoveryConfigurationService.java]: I've included the Javadocs

**Pull request merged**